File: internal/enrichment/relation_inferrer.go
package enrichment

import (
	"strings"

	"github.com/Curt-Park/autology/internal/storage"
)

// InferredRelation represents an inferred relationship between nodes
type InferredRelation struct {
	Source     string
	Target     string
	Type       storage.RelationType
	Confidence float64
	Reasoning  string
}

// InferRelations infers relations for a new node based on existing nodes
func InferRelations(newNode storage.KnowledgeNode, existingNodes []storage.KnowledgeNode) []InferredRelation {
	inferred := make([]InferredRelation, 0)

	for _, existing := range existingNodes {
		// Skip self
		if existing.ID == newNode.ID {
			continue
		}

		// Check various relation signals
		tagOverlap := calculateTagOverlap(newNode.Tags, existing.Tags)
		fileOverlap := calculateFileOverlap(newNode.References, existing.References)
		sameSession := newNode.Session != "" && newNode.Session == existing.Session
		titleSimilarity := calculateTitleSimilarity(newNode.Title, existing.Title)

		// Infer specific relation types
		relation := inferSpecificRelation(newNode, existing, relationSignals{
			tagOverlap:      tagOverlap,
			fileOverlap:     fileOverlap,
			sameSession:     sameSession,
			titleSimilarity: titleSimilarity,
		})

		if relation != nil {
			inferred = append(inferred, *relation)
		}
	}

	// Sort by confidence descending
	for i := 0; i < len(inferred)-1; i++ {
		for j := i + 1; j < len(inferred); j++ {
			if inferred[j].Confidence > inferred[i].Confidence {
				inferred[i], inferred[j] = inferred[j], inferred[i]
			}
		}
	}

	return inferred
}

type relationSignals struct {
	tagOverlap      float64
	fileOverlap     int
	sameSession     bool
	titleSimilarity float64
}

// inferSpecificRelation infers specific relation type based on signals
func inferSpecificRelation(source, target storage.KnowledgeNode, signals relationSignals) *InferredRelation {
	// Rule 1: Decision affects component
	if source.Type == storage.NodeTypeDecision && target.Type == storage.NodeTypeComponent {
		if signals.tagOverlap > 0.3 || signals.fileOverlap > 0 {
			confidence := 0.7 + signals.tagOverlap*0.2
			return &InferredRelation{
				Source:     source.ID,
				Target:     target.ID,
				Type:       storage.RelationTypeAffects,
				Confidence: confidence,
				Reasoning:  formatReasoning("Decision affects component (tag overlap: %.0f%%)", signals.tagOverlap*100),
			}
		}
	}

	// Rule 2: Component uses pattern
	if source.Type == storage.NodeTypeComponent && target.Type == storage.NodeTypePattern {
		if signals.tagOverlap > 0.4 || containsPatternReference(source.Content, target.Title) {
			return &InferredRelation{
				Source:     source.ID,
				Target:     target.ID,
				Type:       storage.RelationTypeImplements,
				Confidence: 0.75,
				Reasoning:  "Component implements pattern",
			}
		}
	}

	// Rule 3: Component uses another component
	if source.Type == storage.NodeTypeComponent && target.Type == storage.NodeTypeComponent {
		if signals.fileOverlap > 0 {
			return &InferredRelation{
				Source:     source.ID,
				Target:     target.ID,
				Type:       storage.RelationTypeUses,
				Confidence: 0.8,
				Reasoning:  formatReasoning("Shared file references: %d file(s)", signals.fileOverlap),
			}
		}
	}

	// Rule 4: Decision supersedes decision
	if source.Type == storage.NodeTypeDecision && target.Type == storage.NodeTypeDecision {
		if containsSupersession(source.Content, target.Title) || containsSupersession(source.Title, target.Title) {
			return &InferredRelation{
				Source:     source.ID,
				Target:     target.ID,
				Type:       storage.RelationTypeSupersedes,
				Confidence: 0.85,
				Reasoning:  "Decision supersedes previous decision",
			}
		}
	}

	// Rule 5: Convention applies to component
	if source.Type == storage.NodeTypeConvention && target.Type == storage.NodeTypeComponent {
		if signals.tagOverlap > 0.3 {
			return &InferredRelation{
				Source:     source.ID,
				Target:     target.ID,
				Type:       storage.RelationTypeRelatesTo,
				Confidence: 0.65,
				Reasoning:  "Convention relates to component",
			}
		}
	}

	// Rule 6: Same session relations (weak)
	if signals.sameSession && signals.tagOverlap > 0.2 {
		return &InferredRelation{
			Source:     source.ID,
			Target:     target.ID,
			Type:       storage.RelationTypeRelatesTo,
			Confidence: 0.6,
			Reasoning:  "Created in same session with shared tags",
		}
	}

	// Rule 7: High tag overlap (generic)
	if signals.tagOverlap > 0.5 {
		confidence := 0.55 + signals.tagOverlap*0.15
		return &InferredRelation{
			Source:     source.ID,
			Target:     target.ID,
			Type:       storage.RelationTypeRelatesTo,
			Confidence: confidence,
			Reasoning:  formatReasoning("High tag overlap: %.0f%%", signals.tagOverlap*100),
		}
	}

	return nil
}

// calculateTagOverlap calculates tag overlap ratio (Jaccard index)
func calculateTagOverlap(tags1, tags2 []string) float64 {
	if len(tags1) == 0 || len(tags2) == 0 {
		return 0
	}

	// Convert to lowercase sets
	set1 := make(map[string]bool)
	for _, tag := range tags1 {
		set1[strings.ToLower(tag)] = true
	}

	set2 := make(map[string]bool)
	for _, tag := range tags2 {
		set2[strings.ToLower(tag)] = true
	}

	// Calculate intersection
	intersection := 0
	for tag := range set1 {
		if set2[tag] {
			intersection++
		}
	}

	// Calculate union
	union := make(map[string]bool)
	for tag := range set1 {
		union[tag] = true
	}
	for tag := range set2 {
		union[tag] = true
	}

	return float64(intersection) / float64(len(union))
}

// calculateFileOverlap calculates number of shared file references
func calculateFileOverlap(refs1, refs2 []string) int {
	if len(refs1) == 0 || len(refs2) == 0 {
		return 0
	}

	set1 := make(map[string]bool)
	for _, ref := range refs1 {
		set1[ref] = true
	}

	overlap := 0
	for _, ref := range refs2 {
		if set1[ref] {
			overlap++
		}
	}

	return overlap
}

// calculateTitleSimilarity calculates title similarity using word overlap
func calculateTitleSimilarity(title1, title2 string) float64 {
	// Extract words > 3 characters
	words1 := make(map[string]bool)
	for _, word := range strings.Fields(strings.ToLower(title1)) {
		if len(word) > 3 {
			words1[word] = true
		}
	}

	words2 := make(map[string]bool)
	for _, word := range strings.Fields(strings.ToLower(title2)) {
		if len(word) > 3 {
			words2[word] = true
		}
	}

	if len(words1) == 0 || len(words2) == 0 {
		return 0
	}

	// Calculate intersection
	intersection := 0
	for word := range words1 {
		if words2[word] {
			intersection++
		}
	}

	// Calculate union
	union := make(map[string]bool)
	for word := range words1 {
		union[word] = true
	}
	for word := range words2 {
		union[word] = true
	}

	return float64(intersection) / float64(len(union))
}

// containsPatternReference checks if content references a pattern name
func containsPatternReference(content, patternName string) bool {
	contentLower := strings.ToLower(content)
	patternLower := strings.ToLower(patternName)

	if strings.Contains(contentLower, patternLower) {
		return true
	}

	// Try without "pattern" suffix
	withoutPattern := strings.Replace(patternLower, " pattern", "", -1)
	if strings.Contains(contentLower, withoutPattern) {
		return true
	}

	// Try with spaces instead of hyphens
	withSpaces := strings.Replace(patternLower, "-", " ", -1)
	return strings.Contains(contentLower, withSpaces)
}

// containsSupersession checks if content indicates supersession
func containsSupersession(content, targetTitle string) bool {
	contentLower := strings.ToLower(content)
	targetLower := strings.ToLower(targetTitle)

	supersessionKeywords := []string{
		"supersedes",
		"replaces",
		"instead of",
		"rather than",
		"deprecates",
		"obsoletes",
		"upgrades from",
	}

	for _, keyword := range supersessionKeywords {
		if strings.Contains(contentLower, keyword) && strings.Contains(contentLower, targetLower) {
			return true
		}
	}

	return false
}

// FilterByConfidence filters inferred relations by confidence threshold
func FilterByConfidence(relations []InferredRelation, minConfidence float64) []InferredRelation {
	filtered := make([]InferredRelation, 0)
	for _, rel := range relations {
		if rel.Confidence >= minConfidence {
			filtered = append(filtered, rel)
		}
	}
	return filtered
}

// GroupByActionResult represents the result of grouping by action
type GroupByActionResult struct {
	AutoCreate []InferredRelation
	Suggest    []InferredRelation
}

// GroupByAction groups inferred relations by action (auto-create vs suggest)
func GroupByAction(relations []InferredRelation, autoCreateThreshold float64) GroupByActionResult {
	autoCreate := make([]InferredRelation, 0)
	suggest := make([]InferredRelation, 0)

	for _, relation := range relations {
		if relation.Confidence >= autoCreateThreshold {
			autoCreate = append(autoCreate, relation)
		} else {
			suggest = append(suggest, relation)
		}
	}

	return GroupByActionResult{
		AutoCreate: autoCreate,
		Suggest:    suggest,
	}
}

// formatReasoning formats reasoning string with printf-style formatting
func formatReasoning(format string, args ...interface{}) string {
	// Simple formatting - handle %.0f%% and %d cases
	result := format
	for _, arg := range args {
		switch v := arg.(type) {
		case float64:
			result = strings.Replace(result, "%.0f%%", formatFloat(v)+"%", 1)
		case int:
			result = strings.Replace(result, "%d", formatInt(v), 1)
		}
	}
	return result
}

func formatFloat(f float64) string {
	return strings.TrimSuffix(strings.TrimSuffix(formatDecimal(f, 0), "."), "0")
}

func formatInt(i int) string {
	if i == 0 {
		return "0"
	}
	result := ""
	for i > 0 {
		result = string(rune('0'+i%10)) + result
		i /= 10
	}
	return result
}

func formatDecimal(f float64, decimals int) string {
	// Simple float to string conversion
	i := int(f)
	result := formatInt(i)
	if decimals > 0 {
		frac := f - float64(i)
		result += "."
		for j := 0; j < decimals; j++ {
			frac *= 10
			result += string(rune('0' + int(frac)%10))
		}
	}
	return result
}

File: packages/mcp-server/src/enrichment/relation-inferrer.ts
/**
 * Automatic relation inference between nodes
 * Discovers potential relationships based on content analysis
 */
import type { KnowledgeNode, RelationType } from '../storage/types.js';

export interface InferredRelation {
  readonly source: string;
  readonly target: string;
  readonly type: RelationType;
  readonly confidence: number;
  readonly reasoning: string;
}

/**
 * Infer relations for a new node based on existing nodes
 */
export function inferRelations(
  newNode: KnowledgeNode,
  existingNodes: ReadonlyArray<KnowledgeNode>
): ReadonlyArray<InferredRelation> {
  const inferred: InferredRelation[] = [];

  for (const existing of existingNodes) {
    // Skip self
    if (existing.id === newNode.id) continue;

    // Check various relation signals
    const tagOverlap = calculateTagOverlap(newNode.tags, existing.tags);
    const fileOverlap = calculateFileOverlap(newNode.references, existing.references);
    const sameSession = !!(newNode.session && newNode.session === existing.session && newNode.session.length > 0);
    const titleSimilarity = calculateTitleSimilarity(newNode.title, existing.title);

    // Infer specific relation types
    const relation = inferSpecificRelation(newNode, existing, {
      tagOverlap,
      fileOverlap,
      sameSession,
      titleSimilarity
    });

    if (relation) {
      inferred.push(relation);
    }
  }

  // Sort by confidence descending
  inferred.sort((a, b) => b.confidence - a.confidence);

  return inferred;
}

interface RelationSignals {
  readonly tagOverlap: number;
  readonly fileOverlap: number;
  readonly sameSession: boolean;
  readonly titleSimilarity: number;
}

/**
 * Infer specific relation type based on signals
 */
function inferSpecificRelation(
  source: KnowledgeNode,
  target: KnowledgeNode,
  signals: RelationSignals
): InferredRelation | null {
  // Rule 1: Decision affects component
  if (source.type === 'decision' && target.type === 'component') {
    if (signals.tagOverlap > 0.3 || signals.fileOverlap > 0) {
      return {
        source: source.id,
        target: target.id,
        type: 'affects',
        confidence: 0.7 + signals.tagOverlap * 0.2,
        reasoning: `Decision affects component (tag overlap: ${(signals.tagOverlap * 100).toFixed(0)}%)`
      };
    }
  }

  // Rule 2: Component uses pattern
  if (source.type === 'component' && target.type === 'pattern') {
    if (signals.tagOverlap > 0.4 || containsPatternReference(source.content, target.title)) {
      return {
        source: source.id,
        target: target.id,
        type: 'implements',
        confidence: 0.75,
        reasoning: 'Component implements pattern'
      };
    }
  }

  // Rule 3: Component uses another component
  if (source.type === 'component' && target.type === 'component') {
    if (signals.fileOverlap > 0) {
      return {
        source: source.id,
        target: target.id,
        type: 'uses',
        confidence: 0.8,
        reasoning: `Shared file references: ${signals.fileOverlap} file(s)`
      };
    }
  }

  // Rule 4: Decision supersedes decision
  if (source.type === 'decision' && target.type === 'decision') {
    if (containsSupersession(source.content, target.title) || containsSupersession(source.title, target.title)) {
      return {
        source: source.id,
        target: target.id,
        type: 'supersedes',
        confidence: 0.85,
        reasoning: 'Decision supersedes previous decision'
      };
    }
  }

  // Rule 5: Convention applies to component
  if (source.type === 'convention' && target.type === 'component') {
    if (signals.tagOverlap > 0.3) {
      return {
        source: source.id,
        target: target.id,
        type: 'relates_to',
        confidence: 0.65,
        reasoning: 'Convention relates to component'
      };
    }
  }

  // Rule 6: Same session relations (weak)
  if (signals.sameSession && signals.tagOverlap > 0.2) {
    return {
      source: source.id,
      target: target.id,
      type: 'relates_to',
      confidence: 0.6,
      reasoning: 'Created in same session with shared tags'
    };
  }

  // Rule 7: High tag overlap (generic)
  if (signals.tagOverlap > 0.5) {
    return {
      source: source.id,
      target: target.id,
      type: 'relates_to',
      confidence: 0.55 + signals.tagOverlap * 0.15,
      reasoning: `High tag overlap: ${(signals.tagOverlap * 100).toFixed(0)}%`
    };
  }

  return null;
}

/**
 * Calculate tag overlap ratio
 */
function calculateTagOverlap(
  tags1: ReadonlyArray<string>,
  tags2: ReadonlyArray<string>
): number {
  if (tags1.length === 0 || tags2.length === 0) return 0;

  const set1 = new Set(tags1.map(t => t.toLowerCase()));
  const set2 = new Set(tags2.map(t => t.toLowerCase()));

  const intersection = new Set([...set1].filter(t => set2.has(t)));
  const union = new Set([...set1, ...set2]);

  return intersection.size / union.size;
}

/**
 * Calculate file reference overlap
 */
function calculateFileOverlap(
  refs1: ReadonlyArray<string>,
  refs2: ReadonlyArray<string>
): number {
  if (refs1.length === 0 || refs2.length === 0) return 0;

  const set1 = new Set(refs1);
  const set2 = new Set(refs2);

  return [...set1].filter(r => set2.has(r)).length;
}

/**
 * Calculate title similarity (simple word overlap)
 */
function calculateTitleSimilarity(title1: string, title2: string): number {
  const words1 = new Set(title1.toLowerCase().split(/\s+/).filter(w => w.length > 3));
  const words2 = new Set(title2.toLowerCase().split(/\s+/).filter(w => w.length > 3));

  if (words1.size === 0 || words2.size === 0) return 0;

  const intersection = new Set([...words1].filter(w => words2.has(w)));
  const union = new Set([...words1, ...words2]);

  return intersection.size / union.size;
}

/**
 * Check if content references a pattern name
 */
function containsPatternReference(content: string, patternName: string): boolean {
  const contentLower = content.toLowerCase();
  const patternLower = patternName.toLowerCase();

  return contentLower.includes(patternLower) ||
         contentLower.includes(patternLower.replace(' pattern', '')) ||
         contentLower.includes(patternLower.replace('-', ' '));
}

/**
 * Check if content indicates supersession
 */
function containsSupersession(content: string, targetTitle: string): boolean {
  const contentLower = content.toLowerCase();
  const targetLower = targetTitle.toLowerCase();

  const supersessionKeywords = [
    'supersedes', 'replaces', 'instead of', 'rather than',
    'deprecates', 'obsoletes', 'upgrades from'
  ];

  for (const keyword of supersessionKeywords) {
    if (contentLower.includes(keyword) && contentLower.includes(targetLower)) {
      return true;
    }
  }

  return false;
}

/**
 * Filter inferred relations by confidence threshold
 */
export function filterByConfidence(
  relations: ReadonlyArray<InferredRelation>,
  minConfidence: number
): ReadonlyArray<InferredRelation> {
  return relations.filter(r => r.confidence >= minConfidence);
}

/**
 * Group inferred relations by action
 */
export function groupByAction(
  relations: ReadonlyArray<InferredRelation>,
  autoCreateThreshold: number = 0.7
): {
  readonly autoCreate: ReadonlyArray<InferredRelation>;
  readonly suggest: ReadonlyArray<InferredRelation>;
} {
  const autoCreate: InferredRelation[] = [];
  const suggest: InferredRelation[] = [];

  for (const relation of relations) {
    if (relation.confidence >= autoCreateThreshold) {
      autoCreate.push(relation);
    } else {
      suggest.push(relation);
    }
  }

  return { autoCreate, suggest };
}
